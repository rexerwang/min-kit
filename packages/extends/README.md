# @min-kit/extends

extends apis of Taro

go [here](https://github.com/rexerwang/min-kit/tree/main/packages/example/src/pkg-extends/pages) for more usage examples.
