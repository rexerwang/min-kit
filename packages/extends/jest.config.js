module.exports = require('@min-kit/jest/config')({
  collectCoverageFrom: [
    // No need to test
    '!src/{current,document}.ts',
  ],
})
