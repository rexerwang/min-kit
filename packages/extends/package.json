{
  "name": "@min-kit/extends",
  "version": "0.5.10",
  "description": "Tarojs apis extends for miniapp",
  "author": "rexerwang",
  "license": "MIT",
  "repository": "rexerwang/min-kit",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "keywords": [
    "min-kit",
    "extends",
    "apis",
    "taro",
    "miniapp",
    "miniprogram"
  ],
  "scripts": {
    "build": "unbuild",
    "test": "jest"
  },
  "dependencies": {
    "@min-kit/shared": "workspace:*"
  },
  "peerDependencies": {
    "@tarojs/runtime": ">=3.6",
    "@tarojs/taro": ">=3.6",
    "react": ">=17"
  },
  "devDependencies": {
    "@min-kit/jest": "workspace:*"
  }
}
