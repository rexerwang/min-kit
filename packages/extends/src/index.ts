export * from './config'
export * from './copy'
export * from './createRewardedVideoAd'
export { default as Current } from './current'
export * from './document'
export * from './go'
export * from './logger'
export * from './request'
export * from './scene'
export * from './shareMenu'
export * from './storage'
export * from './subscription'
export * from './systemInfo'
export * from './toast'
