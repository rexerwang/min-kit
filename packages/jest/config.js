/**
 * @typedef {import('jest').Config Config}
 */

const defineJestConfig = require('./config/taro-jest/index').default

/** @type {Config} */
const baseConfig = defineJestConfig({
  globalSetup: require.resolve('./config/globalSetup.js'),
  setupFilesAfterEnv: [require.resolve('./config/setupAfterEnv.js')],
  testEnvironment: '@happy-dom/jest-environment', // 'jsdom'
  testMatch: ['<rootDir>/src/__tests__/**/*.test.ts?(x)'],
  collectCoverageFrom: ['<rootDir>/src/**/*.ts?(x)', '!**/__tests__/**'],
  coverageReporters: process.env.CI ? ['json'] : ['html', 'text'],
  passWithNoTests: true,
})

/**
 * defineJestConfig
 * @param {Partial<Config>} [config]
 * @returns {Config}
 */
module.exports = (config = {}) => {
  let { setupFilesAfterEnv, collectCoverageFrom, ...configs } = baseConfig

  if (Array.isArray(config.setupFilesAfterEnv)) {
    setupFilesAfterEnv = setupFilesAfterEnv.concat(config.setupFilesAfterEnv)
  }

  if (Array.isArray(config.collectCoverageFrom)) {
    collectCoverageFrom = [collectCoverageFrom[0], ...config.collectCoverageFrom, collectCoverageFrom[1]]
  }

  return {
    ...configs,
    ...config,
    setupFilesAfterEnv,
    collectCoverageFrom,
  }
}
