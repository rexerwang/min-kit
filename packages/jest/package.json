{
  "name": "@min-kit/jest",
  "version": "0.5.11",
  "private": true,
  "description": "jest config & utils for internal",
  "author": "rexerwang",
  "license": "MIT",
  "repository": "rexerwang/min-kit",
  "exports": {
    ".": "./utils.ts",
    "./config": "./config.js"
  },
  "devDependencies": {
    "@happy-dom/jest-environment": "^14.7.1",
    "@tarojs/test-utils-react": "^0.1.1",
    "@testing-library/jest-dom": "^6.4.8",
    "@testing-library/react": "^16.0.1"
  }
}
