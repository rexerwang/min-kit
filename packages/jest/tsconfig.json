{
  "extends": "../../tsconfig.root.json",
  "include": ["utils.ts"]
}
