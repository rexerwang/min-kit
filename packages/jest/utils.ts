import TestUtils from '@tarojs/test-utils-react'

export const createTaroTestUtils = () => {
  const testUtils = new TestUtils()
  return Object.assign(testUtils, { unmount: testUtils.unmout })
}

export const spyOnConsole = (methods = ['error', 'warn', 'info', 'debug']) =>
  methods.map((method: any) => {
    const spy = jest.spyOn(console, method)
    spy.mockImplementation(jest.fn())
    return spy
  })

export * from '@testing-library/react'
