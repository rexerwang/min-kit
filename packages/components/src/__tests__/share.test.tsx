import { Current, document, logger } from '@min-kit/extends'
import { render, renderHook } from '@min-kit/jest'
import Taro from '@tarojs/taro'

describe('share', () => {
  const useShareTimelineSpy = jest.spyOn(Taro, 'useShareTimeline')
  const useShareAppMessageSpy = jest.spyOn(Taro, 'useShareAppMessage')
  const errorSpy = jest.spyOn(logger, 'error').mockImplementation(jest.fn())

  beforeEach(() => {
    useShareTimelineSpy.mockImplementation(jest.fn())
    useShareAppMessageSpy.mockImplementation(jest.fn())
  })

  afterEach(() => {
    useShareTimelineSpy.mockReset()
    useShareAppMessageSpy.mockReset()
    errorSpy.mockClear()
  })

  it('should shareAppMessage via ShareButton', async () => {
    const messageFake = { title: 'test' }
    let message: any = null

    useShareAppMessageSpy.mockImplementation((cb) => {
      message = cb({ from: 'button', target: { id: 'test_id' } })
    })

    jest
      .spyOn(document, 'getElementById')
      .mockImplementation(jest.fn().mockReturnValue({ props: { openType: 'share', message: messageFake } }))

    const { ShareButton } = await import('../share/button')
    const { useShareMessage } = await import('../share/useShareMessage')

    const Index = () => {
      useShareMessage()
      return (
        <ShareButton message={messageFake} onShare={jest.fn()}>
          ShareButton
        </ShareButton>
      )
    }

    expect(render(<Index />).asFragment()).toMatchSnapshot()
    expect(message).toEqual(messageFake)
  })

  it('should shareAppMessage with default message via ShareButton when not found button element', async () => {
    const messageFake = { title: 'test' }
    let message: any = null

    useShareAppMessageSpy.mockImplementation((cb) => {
      message = cb({ from: 'button', target: { id: 'test_id' } })
    })

    const { ShareButton } = await import('../share/button')
    const { useShareMessage } = await import('../share/useShareMessage')

    const Index = () => {
      useShareMessage()
      return (
        <ShareButton message={messageFake} onShare={jest.fn()}>
          ShareButton
        </ShareButton>
      )
    }

    render(<Index />)
    expect(message).toEqual({})
  })

  it('should shareAppMessage via menu', async () => {
    let message: any = null
    useShareAppMessageSpy.mockImplementation((cb) => {
      message = cb({ from: 'menu' })
    })

    const { useShareMessage } = await import('../share/index')

    const { rerender } = renderHook(useShareMessage, { initialProps: { message: { title: 'test' } } })

    expect(message).toEqual({ title: 'test' })

    rerender({})
    expect(message).toEqual({})
  })

  it('should shareAppMessage with default message via menu when useShareAppMessage has no result', async () => {
    let message: any = null
    useShareAppMessageSpy.mockImplementation((cb) => {
      message = cb({})
    })

    const { useShareMessage } = await import('../share/index')

    renderHook(useShareMessage, { initialProps: { message: { title: 'test' } } })
    expect(message).toEqual({})
  })

  it('should shareTimeline via menu', async () => {
    let message: any = null

    useShareTimelineSpy.mockImplementation((cb) => {
      message = cb()
    })

    const { useShareMessage } = await import('../share/index')

    const { rerender } = renderHook(useShareMessage, { initialProps: { timeline: { title: 'timeline' } } })
    expect(message).toEqual({ title: 'timeline' })

    rerender({ message: { title: 'message' } })
    expect(message).toEqual({ title: 'message' })

    rerender({})
    expect(message).toEqual({})
  })

  it('should output error message when `enableShareAppMessage = false`', async () => {
    jest.replaceProperty(Current, 'page', { config: {} } as any)

    const { useShareMessage } = await import('../share/index')
    renderHook(useShareMessage)

    // expect(errorSpy.mock.lastCall?.at(1)).toMatch(/unset "enableShareAppMessage"/)
    expect(errorSpy).toHaveBeenCalledTimes(1)
  })
})
