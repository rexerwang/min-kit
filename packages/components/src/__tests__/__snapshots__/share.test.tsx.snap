// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`share should shareAppMessage via ShareButton 1`] = `
<DocumentFragment>
  <taro-button-core
    class=""
    id="taro:r0:"
    open-type="share"
  >
    <!---->
    ShareButton
  </taro-button-core>
</DocumentFragment>
`;
